from sqlalchemy import create_engine, text

database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

connection_string

engine = create_engine(connection_string)

engine

#Initial EDA
query = '''
SELECT *
FROM scooters
LIMIT 10;
'''

with engine.connect() as connection:
    result = connection.execute(text(query))

result.fetchall()

In [4]:
import pandas as pd

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bird';
'''

with engine.connect() as connection:
    bird = pd.read_sql(text(query), con = connection)

bird.head()

bird.to_csv("birddf.csv", index = False)

In [5]:
#read in bird
birddf = pd.read_csv('../data/birddf.csv')

In [6]:
birddf.shape

(12251590, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bolt';
'''

with engine.connect() as connection:
    bolt = pd.read_sql(text(query), con = connection)

bolt.head()

bolt.to_csv("boltdf.csv", index = False)

In [7]:
#Read in bolt
boltdf = pd.read_csv('../data/boltdf.csv')

In [8]:
boltdf.shape

(3477198, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Gotcha';
'''

with engine.connect() as connection:
    gotcha = pd.read_sql(text(query), con = connection)

gotcha.head()

gotcha.to_csv("gotchadf.csv", index = False)

In [9]:
#Read in gotcha
gotchadf = pd.read_csv('../data/gotchadf.csv')

In [10]:
gotchadf.shape

(4679280, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Jump';
'''

with engine.connect() as connection:
    jump = pd.read_sql(text(query), con = connection)

jump.head()

#Saving jump to csv

jump.to_csv("jumpdf.csv", index = False)

In [11]:
#Read in jump
jumpdf = pd.read_csv('../data/jumpdf.csv')

In [12]:
jumpdf.shape

(21835098, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lime';
'''

with engine.connect() as connection:
    lime = pd.read_sql(text(query), con = connection)

lime.head()

lime.to_csv("limedf.csv", index = False)

In [13]:
limedf = pd.read_csv('../data/limedf.csv')

In [14]:
limedf.shape

(16524261, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lyft';
'''

with engine.connect() as connection:
    lyft = pd.read_sql(text(query), con = connection)

lyft.head()

lyft.to_csv("lyftdf.csv", index = False)

In [15]:
lyftdf = pd.read_csv('../data/lyftdf.csv')

In [16]:
lyftdf.shape

(9087043, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Spin';
'''

with engine.connect() as connection:
    spin = pd.read_sql(text(query), con = connection)

spin.head()

spin.to_csv("spindf.csv", index = False)

In [17]:
spindf = pd.read_csv('../data/spindf.csv')

In [18]:
spindf.shape

(5559573, 9)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    tripsdf = pd.read_sql(text(query), con = connection)

tripsdf.head()

tripsdf.to_csv("tripsdf.csv", index = False)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(query), con = connection)

trips.head()

In [19]:
tripsdf = pd.read_csv('../data/tripsdf.csv')

In [20]:
tripsdf.shape

(565522, 16)

In [21]:
tripsdf.info

<bound method DataFrame.info of                    pubtimestamp companyname triprecordnum  \
0       2019-05-01 00:00:55.423        Bird       BRD2134   
1       2019-05-01 00:03:33.147        Lyft          LFT5   
2       2019-05-01 00:05:55.570        Bird       BRD2168   
3       2019-05-01 00:05:55.570        Bird       BRD2166   
4       2019-05-01 00:05:55.570        Bird       BRD2165   
...                         ...         ...           ...   
565517  2019-08-01 04:53:48.000        JUMP          JMP3   
565518  2019-08-01 04:53:48.000        JUMP          JMP2   
565519  2019-08-01 04:53:48.000        JUMP          JMP1   
565520  2019-08-01 07:04:00.000        JUMP         JMP35   
565521  2019-08-01 07:04:00.000        JUMP         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                           

In [22]:
#Concat to create the scooters table
scooters_list = [birddf, boltdf, gotchadf, jumpdf, limedf, lyftdf, spindf]
scootersdf = pd.concat(scooters_list)

In [23]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0  2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1  2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2  2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3  2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4  2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [24]:
scootersdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73414043 entries, 0 to 5559572
Data columns (total 9 columns):
 #   Column       Dtype  
---  ------       -----  
 0   pubdatetime  object 
 1   latitude     float64
 2   longitude    float64
 3   sumdid       object 
 4   sumdtype     object 
 5   chargelevel  float64
 6   sumdgroup    object 
 7   costpermin   float64
 8   companyname  object 
dtypes: float64(4), object(5)
memory usage: 5.5+ GB


## - Are there any null values in any columns in either table?

In [25]:
#Checking for NaN values in the scooters dataframe
scootersdf.isna().any()

pubdatetime    False
latitude       False
longitude      False
sumdid         False
sumdtype       False
chargelevel     True
sumdgroup      False
costpermin     False
companyname    False
dtype: bool

In [26]:
scootersdf.isnull().sum().sum()

770

In [27]:
#Checking for NaN values in the trips dataframe
tripsdf.isna().any()

pubtimestamp      False
companyname       False
triprecordnum     False
sumdid            False
tripduration      False
tripdistance      False
startdate         False
starttime         False
enddate           False
endtime           False
startlatitude     False
startlongitude    False
endlatitude       False
endlongitude      False
triproute         False
create_dt         False
dtype: bool

## - What date range is represented in each of the date columns? Investigate any values that seem odd.

In [28]:
scootersdf['pubdatetime'] = pd.to_datetime(scootersdf['pubdatetime'], dayfirst = False, format = 'mixed')

In [29]:
scootersdf.dtypes

pubdatetime    datetime64[ns]
latitude              float64
longitude             float64
sumdid                 object
sumdtype               object
chargelevel           float64
sumdgroup              object
costpermin            float64
companyname            object
dtype: object

In [30]:
scootersdf['pubdatetime'].max()

Timestamp('2019-07-31 23:59:57')

In [31]:
scootersdf['pubdatetime'].min()

Timestamp('2019-05-01 00:01:41.247000')

In [32]:
tripsdf['pubtimestamp'] = pd.to_datetime(tripsdf['pubtimestamp'], dayfirst = False, format = 'mixed')

In [33]:
tripsdf.dtypes

pubtimestamp      datetime64[ns]
companyname               object
triprecordnum             object
sumdid                    object
tripduration             float64
tripdistance             float64
startdate                 object
starttime                 object
enddate                   object
endtime                   object
startlatitude            float64
startlongitude           float64
endlatitude              float64
endlongitude             float64
triproute                 object
create_dt                 object
dtype: object

#Checking out the pubtimestamp column of the trips table
tripsdf["pubtimestamp"]

#Checking out the create_dt column in the trips df
tripsdf["create_dt"]

In [34]:
tripsdf['pubtimestamp'].min()

Timestamp('2019-05-01 00:00:55.423000')

In [35]:
tripsdf['pubtimestamp'].max()

Timestamp('2019-08-01 07:04:00')

The scooters df date range is '2019-05-01 00:01:41.247000' - '2019-07-31 23:59:57'.
The trips df date range is '2019-05-01 00:00:55.423000' - '2019-08-01 07:04:00'

## - Is time represented with am/pm or using 24 hour values in each of the columns that include time?

Time is represented in 24 hour values in the time columns in both dataframes.

## - What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [36]:
scootersdf.sumdgroup.unique()

array(['scooter', 'Scooter', 'bicycle'], dtype=object)

We don't need the bicycle sumdgroup for this project.

## - What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [37]:
scootersdf['latitude'].min()

0.0

In [38]:
scootersdf['latitude'].max()

3609874.116666

In [39]:
scootersdf['longitude'].min()

-97.443879

In [40]:
scootersdf['longitude'].max()

0.0

In [41]:
tripsdf['startlatitude'].min()

35.8532

In [42]:
tripsdf['startlatitude'].max()

36.300029

In [43]:
tripsdf['startlongitude'].min()

-86.918008

In [44]:
tripsdf['startlongitude'].max()

-86.3662

The minimum latitude shown is "0.0", as well as the maximum longitude, which is surprising. The maximum latitude is '3609874.116666' which is also "surprising" to say the least.

## -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

In [45]:
tripsdf['tripduration'].min()

-19.3582666667

In [46]:
tripsdf['tripduration'].max()

512619.0

In [47]:
tripsdf['tripdistance'].min()

-20324803.8

In [48]:
tripsdf['tripdistance'].max()

31884482.6476

In [49]:
tripsdf.query('`tripduration`== -19.3582666667')

pubtimestamp companyname triprecordnum         sumdid  \
509599 2019-07-19 00:14:02.297        Lyft         LFT21  Powered853770   

        tripduration  tripdistance   startdate        starttime     enddate  \
509599    -19.358267    4540.68256  2019-07-19  00:09:04.506666  2019-07-18   

                endtime  startlatitude  startlongitude  endlatitude  \
509599  23:49:43.013333       36.15505       -86.78482      36.1469   

        endlongitude                                          triproute  \
509599     -86.79355  [(36.15505, -86.78482), (36.15511, -86.78482),...   

                      create_dt  
509599  2019-07-20 10:52:39.737

In [50]:
tripsdf.query('`tripdistance`== -20324803.8')

pubtimestamp companyname triprecordnum        sumdid  \
550038 2019-07-27 23:56:41.940        Bird       BRD5324  PoweredYZL2A   

        tripduration  tripdistance   startdate        starttime     enddate  \
550038          44.0   -20324803.8  2019-07-27  23:14:23.683333  2019-07-27   

                endtime  startlatitude  startlongitude  endlatitude  \
550038  23:58:23.820000        36.1585        -86.7683       36.126   

        endlongitude                                          triproute  \
550038      -86.7892  [(36.158517, -86.768224), (36.158624, -86.7682...   

                      create_dt  
550038  2019-07-28 05:32:05.610

In [51]:
tripsdf.query('`tripdistance`== 31884482.6476')

pubtimestamp companyname triprecordnum       sumdid  \
447799 2019-07-05 01:18:52      Gotcha          GOT9  Powered2460   

        tripduration  tripdistance   startdate starttime     enddate  \
447799           9.0  3.188448e+07  2019-07-05  01:18:56  2019-07-05   

         endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
447799  01:27:56       36.16338      -86.780142     36.16628    -86.778127   

                                      triproute                create_dt  
447799  [('0', '0'), ('36.16628', '-86.77813')]  2019-07-06 07:13:12.210

## - Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?


In [52]:
scootersdf.companyname.unique()

array(['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin'],
      dtype=object)

In [53]:
tripsdf.companyname.unique()

array(['Bird', 'Lyft', 'Lime', 'Bolt Mobility', 'SPIN', 'Gotcha', 'JUMP'],
      dtype=object)

In the trips df, "Bolt" is "Bolt Mobility", "Jump", and "Spin" are also in upper case.

## 1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [54]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0 2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1 2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2 2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3 2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4 2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [55]:
company_scooters = scootersdf[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

In [56]:
company_scooters

companyname  count
2      Gotcha    224
1        Bolt    360
6        Spin    805
3        Jump   1210
5        Lyft   1735
4        Lime   1824
0        Bird   3860

company_scooters_month = scootersdf[['companyname', 'sumdid', 'pubdatetime']].assign(month = scootersdf['pubdatetime'].dt.month)\
                                                        .groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values()